File: itmd-521/week-04/mnmcount.scala
package main.scala.chapter2

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

/** 
 * Usage: MnMcount <mnm_file_dataset>
 */

object MnMcount {
    def main(args: Array[String]) {
        val spark = SparkSession
        .builder
        .appName("MnMCount")
        .getOrCreate()

        if (args.length < 1) {
            print("usage: MnMcount <mnm_file_dataset>")
            sys.exit(1)
        }
        //Get teh M&M data set filename
        val mnmFile = args(0)
        // Read the file into a Spark DataFrame
        val mnmDF = spark.read.format("csv")
        .option("header", "true")
        .option("inferSchema", "true")
        .load(mnmFile)

        //Aggregate counts of all colors and groupBy() State and Color
        // orderBy() in decending order
        val countMnMDF = mnmDF
        .select("State", "Color", "Count")
        .groupBy("State", "Color")
        .sum("Count")
        .orderBy(desc("sum(Count)"))
        //Show the resulting aggregations for all the states and colors
        countMnMDF.show(60)
        println(s"Total Rows = ${countMnMDF.count()}")
        println()
        // Find aggregate counts for California by filtering
        val caCountMnMDF = mnmDF
        .select("State","Color","Count")
        .where(col("State") == "CA")
        .groupBy("State","Color")
        .sum("Count")
        .orderBy(desc("sum(Count)"))
        // show resulting aggs for California
        caCountMnMDF.show(10)
        //stop spark session
        spark.stop()
    }
}

File: itmd-521/labs/module-03/mnmcount.py
import sys

from pyspark.sql import SparkSession

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: mnmcount <file>", file=sys.stderr)
        sys.exit(-1)

    spark = (SparkSession.builder.appName("Python MnMCount").getOrCreate())

    mnm_file = sys.argv[1]

    mnm_df = (spark.read.format("csv").option("header","true").option("inferSchema","true").load(mnm_file))

    count_mnm_df = (mnm_df.select("State", "Color", "Count").groupBy("State", "Color").sum("Count)").orderBy("sum(Count)", ascending=False))

    count_mnm_df.show(n=60, truncate=False)
    print("Total Rows = %d" % (count_mnm_df.count()))

    spark.stop()